docs/index.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "CONNECT" "1" "September 2010" "" ""
.
.SH "NAME"
\fBconnect\fR \- high performance middleware framework for node
.
.P
Connect is a high performance middleware framework for node \fIhttp://nodejs\.org\fR featuring robust middleware for serving static files, advanced routing, cookie and session implementations, error handling and much more\.
.
.SH "Middleware Usage"
Below is an example which shows usage of the \fIlogger\fR middleware bundled with Connect, as well as \fIstaticProvder\fR\.
.
.IP "" 4
.
.nf

var connect = require(\'connect\');

var server = connect\.createServer(
    connect\.logger(),
    connect\.staticProvider(__dirname + \'/public\')
);

server\.listen(3000);
.
.fi
.
.IP "" 0
.
.SH "Middleware Authoring"
Connect middleware is simply a function which accepts the \fIrequest\fR, \fIresponse\fR objects\. Optionally the third parameter \fInext\fR can be used to continue down the middleware stack\. For example below is a middleware layer that simply responds with "hello world"\.
.
.IP "" 4
.
.nf

function helloWorld(req, res) {
    res\.writeHead(200, { \'Content\-Type\': \'text/plain\' });
    res\.end(\'hello world\');
}

connect\.createServer(helloWorld)\.listen(3000);
.
.fi
.
.IP "" 0
.
.P
Lets say we now have some middleware that will require a setup step, this can be done by returning a closure:
.
.IP "" 4
.
.nf

function respond(msg) {
    msg = msg || \'hello world\';
    return function(req, res) {
        res\.writeHead(200, { \'Content\-Type\': \'text/plain\' });
        res\.end(msg);
    }
}

connect\.createServer(respond(\'wahoo\'))\.listen(3000);
.
.fi
.
.IP "" 0
.
.P
To pass control to the next middleware layer, we may call the \fBnext()\fR function with an optional instanceof \fBError\fR\. Middleware with four parameters is an error handling middleware, the \fBerr\fR object can then be logged, used to issue a response, ignored, etc\.
.
.IP "" 4
.
.nf

function fail(req, res, next) {
    // Exceptions thrown will be automatically passed to next()
    // however for custom exceptions / async exceptions you may pass them
    next(new Error(\'something broke!\'));
}

function errorHandler(err, req, res, next) {
    res\.writeHead(500, { \'Content\-Type\': \'text/plain\' });
    res\.end(err\.stack);
}

connect\.createServer(fail, errorHandler)\.listen(3000);
.
.fi
.
.IP "" 0
.
.P
To make your middleware available to others, typically you write a module, and export the function itself:
.
.IP "" 4
.
.nf

  // delay\.js
  module\.exports = function(ms){
      ms = ms || 1000;
      return function(req, res, next){
        setTimeout(next, ms);
      }
  };

 // app\.js
 // delay one second before continuing down the stack
 connect\.createServer(require(\'\./delay\')(1000))\.listen(3000);
.
.fi
.
.IP "" 0
.
.SH "Bundled Middleware"
Connect ships with several helpful middleware modules, the following are currently provided out of the box:
.
.IP "" 4
.
.nf

bodyDecoder      Buffers and parses json and urlencoded request bodies (extenable)
conditionalGet   Provides 304 "Not Modified" support
errorHandler     Handles exceptions thrown, or passed through the stack
gzip             Compresses response bodies with gzip executable
lint             Aids in middleware development
logger           Provides common logger support, and custom log formats
methodOverride   Provides faux HTTP method support by using the "_method" key by default
redirect         Provides req\.redirect() with "magic" urls, ex: req\.redirect("back")
compiler         Supports arbitrary static compilation of files, currently supports less and sass\.
cacheManifest    Provides cache manifest for offline apps
jsonrpc          Provides JSON\-RPC 2\.0 support
staticProvider   Serves static files
router           Provides a feature rich routing API similar to Sinatra and Express
cookieDecoder    Provides cookie parsing support
session          Provides session support
cache            Provides memory caching
repl             Read Evaluate Print Loop attached to "/tmp/connect\.sock" for inspecting live servers
vhost            Virtual host support
.
.fi
.
.IP "" 0
.
.SS "Middleware Documentation"
To view middleware specific documentation execute:
.
.IP "" 4
.
.nf

$ man connect\-MIDDLEWARE
.
.fi
.
.IP "" 0
.
.P
For example:
.
.IP "" 4
.
.nf

$ man connect\-bodyDecoder
.
.fi
.
.IP "" 0



lib/connect/index.js

/*!
 * Ext JS Connect
 * Copyright(c) 2010 Sencha Inc.
 * MIT Licensed
 */

/**
 * Framework version.
 */

exports.version = '0.5.2';

/**
 * Module dependencies.
 */

var fs = require('fs')
  , http = require('http')
  , utils = require('./utils')
  , assert = require('assert');


/**
 * COMPAT: 0.3.x support
 */

try {
  var consts = require('constants');
  Object.keys(consts).forEach(function (key) {
    process[key] = consts[key];
  });
} catch (err) {
  // Ignore
}

/**
 * Default env for those who do not use spark.
 */

process.connectEnv = process.sparkEnv || {
    name: process.env.NODE_ENV || 'development'
  , port: 3000
  , host: null
};

/**
 * Setup the given middleware layer callbacks.
 *
 * Examples:
 *
 *     connect.createServer(
 *        connect.log(),
 *        connect.staticProvider(__dirname + '/public')
 *     );
 *
 *     var server = connect.createServer();
 *     server.use('/', connect.log());
 *     server.use('/', connect.static(__dirname + '/public'));
 *
 * @params {Function} ...
 * @return {Server}
 * @api public
 */

var Server = exports.Server = function Server(layers) {
  this.stack = [];

  // Stack layers
  layers.forEach(function(layer){
    this.use("/", layer);
  }, this);

  // Set up the request handler using the parent's constructor
  http.Server.call(this, this.handle);
};

/**
 * Inherit from `http.Server.prototype`.
 */

Server.prototype.__proto__ = http.Server.prototype;

/**
 * Stack the given middleware `handle` to the given `route`.
 *
 * @param {String|Function} route or handle
 * @param {Function} handle
 * @return {Server}
 * @api public
 */

Server.prototype.use = function(route, handle){
  if (typeof route !== 'string') {
    handle = route;
    route = '/';
  }

  // Allow for more than one handle function to be added at a time
  if (arguments.length > 2) {
    Array.prototype.slice.call(arguments, 1).forEach(function (handle) {
      this.use(route, handle);
    }, this);
    return this;
  }

  // Wrap sub-apps
  if (handle instanceof Server) {
    var server = handle;
    server.route = route;
    handle = function(req, res, next) {
      server.handle(req, res, next);
    };
  }

  // Wrap vanilla http.Server instances too
  // Maybe later we'll handle 404 and 500 responses and pass them to next()
  // But for now, this is the end of the chain
  if (handle instanceof http.Server) {
    handle = handle.listeners('request')[0];
  }

  // Validate the input
  if (!(typeof route === 'string' && typeof handle === 'function')) {
    throw new Error("Each layer must have a route and a handle function");
  }

  // Normalize route to not trail with slash
  if (route[route.length - 1] === '/') {
    route = route.substr(0, route.length - 1);
  }

  // Add the route, handle pair to the stack
  this.stack.push({ route: route, handle: handle });

  // Allow chaining
  return this;
};

/**
 * Listen on the given `port` number,
 * defaults to the current environment's port,
 * or _3000_ when neither are present.
 *
 * @param  {Number} port
 * @param  {String} host
 * @return {Server}
 * @api public
 */

Server.prototype.listen = function(port, host) {
  arguments[0] = port || process.connectEnv.port || 3000;
  arguments[1] = host || process.connectEnv.host;
  http.Server.prototype.listen.apply(this, arguments);
  return this;
};

/**
 * Handle server requests, punting them down
 * the middleware stack.
 *
 * @api private
 */

Server.prototype.handle = function(req, res, outerNext) {
  var writeHead = res.writeHead
    , removed = ""
    , stack = this.stack
    , index = 0;

  function next(err) {
    // Put the prefix back on if it was removed
    req.url = req.originalUrl = removed + req.url;
    removed = "";

    var layer = stack[index++];

    // If we reached the end of the chain...
    if (!layer) {
      // ... and we're part of a bigger chain then forward to it.
      if (outerNext) {
        outerNext(err);
        return;
      }
      // Otherwise send a proper error message to the browser.
      if (err) {
        var msg = process.connectEnv.name === 'production'
          ? 'Internal Server Error'
          : err.stack || err.toString();
        if (process.connectEnv.name !== 'test') {
          console.error(err.stack || err.toString());
        }
        res.writeHead(500, { 'Content-Type': 'text/plain' });
        res.end(msg);
      } else {
        res.writeHead(404, { 'Content-Type': 'text/plain' });
        res.end('Cannot ' + req.method + ' ' + req.url);
      }
      return;
    }

    try {
      var path = utils.pathname(req.url);
      if (path === undefined) path = '/';

      // Skip this layer if the route doesn't match.
      if (path.indexOf(layer.route) !== 0) {
        next(err);
        return;
      }

      var nextChar = path[layer.route.length];
      if (nextChar && nextChar !== '/' && nextChar !== '.') {
        next(err);
        return;
      }

      // Call the layer handler
      // Trim off the part of the url that matches the route
      removed = layer.route;
      req.url = req.url.substr(removed.length);

      // Ensure leading slash
      if (req.url[0] !== "/") {
        req.url = "/" + req.url;
      }

      var arity = layer.handle.length;
      if (err) {
        if (arity === 4) {
          layer.handle(err, req, res, next);
        } else {
          next(err);
        }
      } else if (arity < 4) {
        layer.handle(req, res, next);
      } else {
        next();
      }
    } catch (e) {
      if (e instanceof assert.AssertionError) {
        console.error(e.stack + '\n');
        next(e);
      } else {
        next(e);
      }
    }
  }
  next();
};

/**
 * Shortcut for `new connect.Server([ ... ])`.
 *
 * @param  {Function} ...
 * @param  {Object} env
 * @return {Server}
 * @api public
 */

exports.createServer = function() {
  return new Server(Array.prototype.slice.call(arguments));
};

/**
 * Auto-load getters.
 */

exports.middleware = {};

/**
 * Auto-load bundled middleware with getters.
 */

fs.readdirSync(__dirname + '/middleware').forEach(function(filename){
  if (/\.js$/.test(filename)) {
    var name = filename.substr(0, filename.lastIndexOf('.'));
    Object.defineProperty(exports.middleware, name, { get: function(){
      return require('./middleware/' + name);
    }});
  }
});

/**
 * Expose getters as first-class exports.
 */

exports.__proto__ = exports.middleware;

/**
 * Expose utils.
 */

exports.utils = require('./utils');
